### DISCUSSION TITLE
Python Asyncio: The Complete Guide

### DISCUSSION THREAD

**Discussion by fivdi** [micropython/micropython#10063] [Show and tell]
Here's a link to a fairly extensive guide to asynchronous programming in Python that I find good:

[![Python Asyncio: The Complete Guide](https://superfastpython.com/wp-content/uploads/2022/11/Python-Asyncio-Guide_300.png)](https://superfastpython.com/python-asyncio/)

[Python Asyncio: The Complete Guide](https://superfastpython.com/python-asyncio/) by Jason Brownlee.

It's a good supplement to [Asynchronous programming in MicroPython](https://github.com/peterhinch/micropython-async) by Peter Hinch.

**Comment by beetlegigg** [comment]
Thanks for the link, I've been reading the first few chapters and its very well written.  My raspberry pi will soon be humming away processing many coroutines.. :-)    I've been wondering if / how to leverage my delve into asyncio in micropython guided by the redoubtable Peter Hinch with his most excellent treatise on async in mp. (thanks Peter)

**Comment by peterhinch** [comment]
I would suggest extending this thread to add other asynchronous programming resources. To kick it off, here are some that I've found useful.

[This video](https://www.youtube.com/watch?v=kdzL3r-yJZY) provides a good introduction to asynchronous concepts.  
[Threads are bad](https://glyph.twistedmatrix.com/2014/02/unyielding.html) is an excellent explanation of why you should avoid threading unless you know exactly why you're using it.  

[The red-blue function problem](https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/) is a critique of the asyncio paradigm. It won't help you write MP code, but it is food for thought,

**Comment by aligheshlaghi97** [comment]
Hi everyone and thanks for sharing the resources!

I’ve read `Asynchronous Programming in MicroPython` by Peter Hinch which was quite good, specially if you're into embedded systems (as I was in the past 😅). I also found `Python Asyncio: The Complete Guide` by Jason Brownlee particularly inspiring, it actually inspired me to build my own guide. Other links were also very interesting for me.

I put together [asynchronous-python](https://aligheshlaghi97.github.io/asynchronous-python/), a comprehensive and hands-on guide that covers asynchronous programming in Python from core concepts to practical use cases. It’s compatible with the latest Python versions (3.11 to 3.13) and includes working code snippets with passing unit tests.

### Some of the topics covered include:

* Event loops, coroutines, and `asyncio` APIs
* CPU-bound vs I/O-bound workloads
* Process creation (`fork` vs `spawn`)
* Synchronization primitives like `locks` and `semaphores`
* `asyncio.Queue`, `Future`, callbacks, and error handling
* Async-related improvements in Python 3.11 to 3.13
* `ASGI` vs `WSGI`, and building async web APIs with `Starlette`
* Deep dives into `AnyIO` and `Trio` with structured concurrency
* Task scheduling with `Celery` and `APScheduler`


#### Would appreciate any feedback or comments 😊

**Reply by peterhinch** [reply]
That looks good as a guide to CPython's `asyncio`. From a MicroPython perspective I have a couple of reservations.

Firstly it includes features absent in MicroPython's lightweight `asyncio` implementation: this could be confusing to beginners. Some of the examples require "big iron": for example trying to launch 10,000 tasks on a microcontroller is likely to result in tears before bedtime...

**Reply by aligheshlaghi97** [reply]
Thank you so much for taking the time to review it, I truly appreciate your thoughtful feedback 🙏

You're absolutely right that my guide is focused on `CPython's asyncio` and assumes it's being run in a resource-rich environment.

Also, I’ve learned a lot from your `MicroPython` tutorial and really admire the work you’ve done to bring clarity. It means a lot to get feedback from someone with your depth of experience.

If you don’t mind me asking: do you have any suggestions on how I could better shape or scope this kind of content to be more useful for those coming from the `MicroPython` side? Or you think there should be another chapter for the tutorial to discuss `MicroPython`?

**Reply by peterhinch** [reply]
I'm not sure what to recommend. When I first approached MicroPython `asyncio` I studied the CPython docs. Despite long experience of cooperative scheduling, I found the docs confusing, mainly because `asyncio` had grown by accretion over numerous iterations. I set about writing the [tutorial](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/TUTORIAL.md) to describe only the MicroPython subset; also to cover firmware-specific issues.

You rightly cover features beyond the MP subset, including OS API elements unavailable to a baremetal port (e.g. `fork` and `spawn`). A lot of the tools you describe are beyond the scope of microcontrollers. While this is of great interest, it might confuse a beginner in microcontroller coding.

Your best bet might be a discussion of the MP subset with pointers to the docs.
